#
# OpenSSH 8.2 delivered support for FIDO/U2F hardware authenticators.
# It depends on libfido2 library (and its dependencies) which is currently
# not available in Solaris. Thus, the patch disables FIDO/U2F in Solaris.
#
# As such, this is Solaris specific patch.
#
--- old/sshkey.c	2020-11-19 15:16:11.497051026 +0000
+++ new/sshkey.c	2020-11-19 15:22:03.844421491 +0000
@@ -150,10 +150,12 @@ static const struct keytype keytypes[] =
 	{ "ssh-ed25519", "ED25519", NULL, KEY_ED25519, 0, 0, 0 },
 	{ "ssh-ed25519-cert-v01@openssh.com", "ED25519-CERT", NULL,
 	    KEY_ED25519_CERT, 0, 1, 0 },
+#ifdef ENABLE_SK
 	{ "sk-ssh-ed25519@openssh.com", "ED25519-SK", NULL,
 	    KEY_ED25519_SK, 0, 0, 0 },
 	{ "sk-ssh-ed25519-cert-v01@openssh.com", "ED25519-SK-CERT", NULL,
 	    KEY_ED25519_SK_CERT, 0, 1, 0 },
+#endif /* ENABLE_SK */
 #ifdef WITH_XMSS
 	{ "ssh-xmss@openssh.com", "XMSS", NULL, KEY_XMSS, 0, 0, 0 },
 	{ "ssh-xmss-cert-v01@openssh.com", "XMSS-CERT", NULL,
@@ -173,10 +175,12 @@ static const struct keytype keytypes[] =
 	{ "ecdsa-sha2-nistp521", "ECDSA", NULL,
 	    KEY_ECDSA, NID_secp521r1, 0, 0 },
 #  endif /* OPENSSL_HAS_NISTP521 */
+#  ifdef ENABLE_SK
 	{ "sk-ecdsa-sha2-nistp256@openssh.com", "ECDSA-SK", NULL,
 	    KEY_ECDSA_SK, NID_X9_62_prime256v1, 0, 0 },
 	{ "webauthn-sk-ecdsa-sha2-nistp256@openssh.com", "ECDSA-SK", NULL,
 	    KEY_ECDSA_SK, NID_X9_62_prime256v1, 0, 1 },
+#  endif /* ENABLE_SK */
 # endif /* OPENSSL_HAS_ECC */
 	{ "ssh-rsa-cert-v01@openssh.com", "RSA-CERT", NULL,
 	    KEY_RSA_CERT, 0, 1, 0 },
@@ -195,8 +199,10 @@ static const struct keytype keytypes[] =
 	{ "ecdsa-sha2-nistp521-cert-v01@openssh.com", "ECDSA-CERT", NULL,
 	   KEY_ECDSA_CERT, NID_secp521r1, 1, 0 },
 #  endif /* OPENSSL_HAS_NISTP521 */
+#  ifdef ENABLE_SK
 	{ "sk-ecdsa-sha2-nistp256-cert-v01@openssh.com", "ECDSA-SK-CERT", NULL,
 	    KEY_ECDSA_SK_CERT, NID_X9_62_prime256v1, 1, 0 },
+#  endif /* ENABLE_SK */
 # endif /* OPENSSL_HAS_ECC */
 #endif /* WITH_OPENSSL */
 	{ "null", "null", NULL, KEY_NULL, 0, 0, 0 },
--- old/ssh-keygen.1	2020-11-19 15:16:11.503596093 +0000
+++ new/ssh-keygen.1	2020-11-19 15:22:03.850115492 +0000
@@ -51,7 +51,7 @@
 .Op Fl m Ar format
 .Op Fl N Ar new_passphrase
 .Op Fl O Ar option
-.Op Fl t Cm dsa | ecdsa | ecdsa-sk | ed25519 | ed25519-sk | rsa
+.Op Fl t Cm dsa | ecdsa | ed25519 | rsa
 .Op Fl w Ar provider
 .Nm ssh-keygen
 .Fl p
@@ -291,7 +291,7 @@ flag determines the key length by select
 curve sizes: 256, 384 or 521 bits.
 Attempting to use bit lengths other than these three values for ECDSA keys
 will fail.
-ECDSA-SK, Ed25519 and Ed25519-SK keys have a fixed length and the
+Ed25519 key has a fixed length and the
 .Fl b
 flag will be ignored.
 .It Fl C Ar comment
@@ -574,14 +574,12 @@ by key ID or serial number.
 See the
 .Sx KEY REVOCATION LISTS
 section for details.
-.It Fl t Cm dsa | ecdsa | ecdsa-sk | ed25519 | ed25519-sk | rsa
+.It Fl t Cm dsa | ecdsa | ed25519 | rsa
 Specifies the type of key to create.
 The possible values are
 .Dq dsa ,
 .Dq ecdsa ,
-.Dq ecdsa-sk ,
 .Dq ed25519 ,
-.Dq ed25519-sk ,
 or
 .Dq rsa .
 .Pp
@@ -977,7 +975,8 @@ of user presence (e.g. by having the use
 This option only makes sense for the FIDO authenticator algorithms
 .Cm ecdsa-sk
 and
-.Cm ed25519-sk .
+.Cm ed25519-sk
+which are currently not supported on Solaris.
 .Pp
 .It Ic source-address Ns = Ns Ar address_list
 Restrict the source addresses from which the certificate is considered valid.
@@ -1151,18 +1150,16 @@ user2@example.com namespaces="file" ssh-
 .It Ev SSH_SK_PROVIDER
 Specifies a path to a library that will be used when loading any
 FIDO authenticator-hosted keys, overriding the default of using
-the built-in USB HID support.
+the built-in USB HID support. Not used on Solaris.
 .El
 .Sh FILES
 .Bl -tag -width Ds -compact
 .It Pa ~/.ssh/id_dsa
 .It Pa ~/.ssh/id_ecdsa
-.It Pa ~/.ssh/id_ecdsa_sk
 .It Pa ~/.ssh/id_ed25519
-.It Pa ~/.ssh/id_ed25519_sk
 .It Pa ~/.ssh/id_rsa
-Contains the DSA, ECDSA, authenticator-hosted ECDSA, Ed25519,
-authenticator-hosted Ed25519 or RSA authentication identity of the user.
+Contains the DSA, ECDSA, Ed25519
+or RSA authentication identity of the user.
 This file should not be readable by anyone but the user.
 It is possible to
 specify a passphrase when generating the key; that passphrase will be
@@ -1175,12 +1172,10 @@ will read this file when a login attempt
 .Pp
 .It Pa ~/.ssh/id_dsa.pub
 .It Pa ~/.ssh/id_ecdsa.pub
-.It Pa ~/.ssh/id_ecdsa_sk.pub
 .It Pa ~/.ssh/id_ed25519.pub
-.It Pa ~/.ssh/id_ed25519_sk.pub
 .It Pa ~/.ssh/id_rsa.pub
-Contains the DSA, ECDSA, authenticator-hosted ECDSA, Ed25519,
-authenticator-hosted Ed25519 or RSA public key for authentication.
+Contains the DSA, ECDSA, Ed25519
+or RSA public key for authentication.
 The contents of this file should be added to
 .Pa ~/.ssh/authorized_keys
 on all machines
--- old/ssh_config.5	2020-11-19 15:16:11.509246658 +0000
+++ new/ssh_config.5	2020-11-19 15:22:03.855660245 +0000
@@ -901,15 +901,12 @@ The default for this option is:
 ecdsa-sha2-nistp256-cert-v01@openssh.com,
 ecdsa-sha2-nistp384-cert-v01@openssh.com,
 ecdsa-sha2-nistp521-cert-v01@openssh.com,
-sk-ecdsa-sha2-nistp256-cert-v01@openssh.com,
 ssh-ed25519-cert-v01@openssh.com,
-sk-ssh-ed25519-cert-v01@openssh.com,
 rsa-sha2-512-cert-v01@openssh.com,
 rsa-sha2-256-cert-v01@openssh.com,
 ssh-rsa-cert-v01@openssh.com,
 ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,
-sk-ecdsa-sha2-nistp256@openssh.com,
-ssh-ed25519,sk-ssh-ed25519@openssh.com,
+ssh-ed25519,
 rsa-sha2-512,rsa-sha2-256,ssh-rsa
 .Ed
 .Pp
@@ -938,15 +935,12 @@ The default for this option is:
 ecdsa-sha2-nistp256-cert-v01@openssh.com,
 ecdsa-sha2-nistp384-cert-v01@openssh.com,
 ecdsa-sha2-nistp521-cert-v01@openssh.com,
-sk-ecdsa-sha2-nistp256-cert-v01@openssh.com,
 ssh-ed25519-cert-v01@openssh.com,
-sk-ssh-ed25519-cert-v01@openssh.com,
 rsa-sha2-512-cert-v01@openssh.com,
 rsa-sha2-256-cert-v01@openssh.com,
 ssh-rsa-cert-v01@openssh.com,
 ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,
-sk-ecdsa-sha2-nistp256@openssh.com,
-ssh-ed25519,sk-ssh-ed25519@openssh.com,
+ssh-ed25519,
 rsa-sha2-512,rsa-sha2-256,ssh-rsa
 .Ed
 .Pp
@@ -1447,15 +1441,12 @@ The default for this option is:
 ecdsa-sha2-nistp256-cert-v01@openssh.com,
 ecdsa-sha2-nistp384-cert-v01@openssh.com,
 ecdsa-sha2-nistp521-cert-v01@openssh.com,
-sk-ecdsa-sha2-nistp256-cert-v01@openssh.com,
 ssh-ed25519-cert-v01@openssh.com,
-sk-ssh-ed25519-cert-v01@openssh.com,
 rsa-sha2-512-cert-v01@openssh.com,
 rsa-sha2-256-cert-v01@openssh.com,
 ssh-rsa-cert-v01@openssh.com,
 ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,
-sk-ecdsa-sha2-nistp256@openssh.com,
-ssh-ed25519,sk-ssh-ed25519@openssh.com,
+ssh-ed25519,
 rsa-sha2-512,rsa-sha2-256,ssh-rsa
 .Ed
 .Pp
--- old/sshd_config.5	2020-11-19 15:16:11.515006805 +0000
+++ new/sshd_config.5	2020-11-19 15:22:03.860985243 +0000
@@ -729,15 +729,12 @@ The default for this option is:
 ecdsa-sha2-nistp256-cert-v01@openssh.com,
 ecdsa-sha2-nistp384-cert-v01@openssh.com,
 ecdsa-sha2-nistp521-cert-v01@openssh.com,
-sk-ecdsa-sha2-nistp256-cert-v01@openssh.com,
 ssh-ed25519-cert-v01@openssh.com,
-sk-ssh-ed25519-cert-v01@openssh.com,
 rsa-sha2-512-cert-v01@openssh.com,
 rsa-sha2-256-cert-v01@openssh.com,
 ssh-rsa-cert-v01@openssh.com,
 ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,
-sk-ecdsa-sha2-nistp256@openssh.com,
-ssh-ed25519,sk-ssh-ed25519@openssh.com,
+ssh-ed25519,
 rsa-sha2-512,rsa-sha2-256,ssh-rsa
 .Ed
 .Pp
@@ -812,15 +809,12 @@ The default for this option is:
 ecdsa-sha2-nistp256-cert-v01@openssh.com,
 ecdsa-sha2-nistp384-cert-v01@openssh.com,
 ecdsa-sha2-nistp521-cert-v01@openssh.com,
-sk-ecdsa-sha2-nistp256-cert-v01@openssh.com,
 ssh-ed25519-cert-v01@openssh.com,
-sk-ssh-ed25519-cert-v01@openssh.com,
 rsa-sha2-512-cert-v01@openssh.com,
 rsa-sha2-256-cert-v01@openssh.com,
 ssh-rsa-cert-v01@openssh.com,
 ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,
-sk-ecdsa-sha2-nistp256@openssh.com,
-ssh-ed25519,sk-ssh-ed25519@openssh.com,
+ssh-ed25519,
 rsa-sha2-512,rsa-sha2-256,ssh-rsa
 .Ed
 .Pp
@@ -1528,15 +1522,12 @@ The default for this option is:
 ecdsa-sha2-nistp256-cert-v01@openssh.com,
 ecdsa-sha2-nistp384-cert-v01@openssh.com,
 ecdsa-sha2-nistp521-cert-v01@openssh.com,
-sk-ecdsa-sha2-nistp256-cert-v01@openssh.com,
 ssh-ed25519-cert-v01@openssh.com,
-sk-ssh-ed25519-cert-v01@openssh.com,
 rsa-sha2-512-cert-v01@openssh.com,
 rsa-sha2-256-cert-v01@openssh.com,
 ssh-rsa-cert-v01@openssh.com,
 ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,
-sk-ecdsa-sha2-nistp256@openssh.com,
-ssh-ed25519,sk-ssh-ed25519@openssh.com,
+ssh-ed25519,
 rsa-sha2-512,rsa-sha2-256,ssh-rsa
 .Ed
 .Pp
--- old/ssh.1	2020-11-19 15:16:11.520321689 +0000
+++ new/ssh.1	2020-11-19 15:22:03.866267040 +0000
@@ -289,9 +289,7 @@ public key authentication is read.
 The default is
 .Pa ~/.ssh/id_dsa ,
 .Pa ~/.ssh/id_ecdsa ,
-.Pa ~/.ssh/id_ecdsa_sk ,
 .Pa ~/.ssh/id_ed25519 ,
-.Pa ~/.ssh/id_ed25519_sk
 and
 .Pa ~/.ssh/id_rsa .
 Identity files may also be specified on
@@ -924,12 +922,8 @@ This stores the private key in
 (DSA),
 .Pa ~/.ssh/id_ecdsa
 (ECDSA),
-.Pa ~/.ssh/id_ecdsa_sk
-(authenticator-hosted ECDSA),
 .Pa ~/.ssh/id_ed25519
 (Ed25519),
-.Pa ~/.ssh/id_ed25519_sk
-(authenticator-hosted Ed25519),
 or
 .Pa ~/.ssh/id_rsa
 (RSA)
@@ -938,12 +932,8 @@ and stores the public key in
 (DSA),
 .Pa ~/.ssh/id_ecdsa.pub
 (ECDSA),
-.Pa ~/.ssh/id_ecdsa_sk.pub
-(authenticator-hosted ECDSA),
 .Pa ~/.ssh/id_ed25519.pub
 (Ed25519),
-.Pa ~/.ssh/id_ed25519_sk.pub
-(authenticator-hosted Ed25519),
 or
 .Pa ~/.ssh/id_rsa.pub
 (RSA)
@@ -1539,9 +1529,7 @@ above.
 .Pp
 .It Pa ~/.ssh/id_dsa
 .It Pa ~/.ssh/id_ecdsa
-.It Pa ~/.ssh/id_ecdsa_sk
 .It Pa ~/.ssh/id_ed25519
-.It Pa ~/.ssh/id_ed25519_sk
 .It Pa ~/.ssh/id_rsa
 Contains the private key for authentication.
 These files
@@ -1555,9 +1543,7 @@ sensitive part of this file using AES-12
 .Pp
 .It Pa ~/.ssh/id_dsa.pub
 .It Pa ~/.ssh/id_ecdsa.pub
-.It Pa ~/.ssh/id_ecdsa_sk.pub
 .It Pa ~/.ssh/id_ed25519.pub
-.It Pa ~/.ssh/id_ed25519_sk.pub
 .It Pa ~/.ssh/id_rsa.pub
 Contains the public key for authentication.
 These files are not
--- old/sshd.8	2020-11-19 15:16:11.525392742 +0000
+++ new/sshd.8	2020-11-19 15:22:03.871720235 +0000
@@ -433,16 +433,12 @@ The supported key types are:
 .Pp
 .Bl -item -compact -offset indent
 .It
-sk-ecdsa-sha2-nistp256@openssh.com
-.It
 ecdsa-sha2-nistp256
 .It
 ecdsa-sha2-nistp384
 .It
 ecdsa-sha2-nistp521
 .It
-sk-ssh-ed25519@openssh.com
-.It
 ssh-ed25519
 .It
 ssh-dss
@@ -459,9 +455,7 @@ Note that lines in this file can be seve
 You don't want to type them in; instead, copy the
 .Pa id_dsa.pub ,
 .Pa id_ecdsa.pub ,
-.Pa id_ecdsa_sk.pub ,
 .Pa id_ed25519.pub ,
-.Pa id_ed25519_sk.pub ,
 or the
 .Pa id_rsa.pub
 file and edit it.
@@ -640,7 +634,8 @@ the user, e.g. via a PIN.
 This option only makes sense for the FIDO authenticator algorithms
 .Cm ecdsa-sk
 and
-.Cm ed25519-sk .
+.Cm ed25519-sk
+which are not supported on Solaris.
 .It Cm restrict
 Enable all restrictions, i.e. disable port, agent and X11 forwarding,
 as well as disabling PTY allocation
@@ -684,8 +679,6 @@ restrict,command="uptime" ssh-rsa AAAA1C
 user@example.net
 restrict,pty,command="nethack" ssh-rsa AAAA1f8...IrrC5==
 user@example.net
-no-touch-required sk-ecdsa-sha2-nistp256@openssh.com AAAAInN...Ko==
-user@example.net
 .Ed
 .Sh SSH_KNOWN_HOSTS FILE FORMAT
 The
